### PR TITLE
Ensure _non_special_token always gives us a token

### LINKED ISSUE #119: Don't discard non-special tokens in macro invocations
```
vec![Foo::bar()];
baz![?];
```

The `::` and `?` are not included in the syntax tree. I think this is because `_non_special_token` is matching these with a regular expression, unlike e.g. `as`.

https://github.com/tree-sitter/tree-sitter-rust/blob/cc7bdd3e6d14677e8aa77da64f6a3f57b6f8b00a/grammar.js#L193-L198

This is causing issues for [difftastic](https://github.com/wilfred/difftastic), where `Foo::bar()` preserves the `::` but `vec![Foo::bar()]` does not.

### PR DESCRIPTION
**Pull request by Wilfred**
Previously, this parser would discard tokens inside macro invocation. As suggested in https://github.com/tree-sitter/tree-sitter/issues/1156, use a named rule to ensure a token is returned.

Closes #119

### CODE REVIEW

**Comment by aryx** [comment]
@dcreager is there someone at github in charge of maintaining tree-sitter-rust? I see a few PRs without answers.

File: corpus/macros.txt [hunk 1]
```diff
@@ -1,25 +1,41 @@
-============================================
+================================================================================
```
Comment (aryx) [inline corpus/macros.txt:1]: Could you remove from this PR all those esthetic changes and just focus on the change mentioned in the PR?

File: corpus/macros.txt [hunk 2]
```diff
@@ -31,56 +47,81 @@ a!('lifetime)
...
 }
 
----
+--------------------------------------------------------------------------------
 
 (source_file
   (macro_invocation
     (identifier)
-    (token_tree (line_comment) (block_comment))))
+    (token_tree
```
Comment (aryx) [inline corpus/macros.txt:118]: Again, could you remove all those indentation changes that makes it more difficult to review this PR.
Remove all unrelated changes.

File: grammar.js [hunk 3]
```diff
@@ -190,10 +190,12 @@ module.exports = grammar({
       '$', '(', repeat($._tokens), ')', optional(/[^+*?]+/), choice('+', '*', '?')
     ),
 
+    non_special_punctuation: $ => /[/_\-=->,;:::!=?.@*&#%^+<>|~]+/,
```
Comment (aryx) [inline grammar.js:193]: Thanks, this will be useful also in semgrep where we need those named rules.

**Review by aryx — CHANGES_REQUESTED** [review]

**Comment by Wilfred** [comment]
@aryx Looks like the corpus text files were generated with an older version of tree-sitter, but I'm not sure which. I've manually merged my changes in.

File: grammar.js [hunk 4]
```diff
@@ -190,10 +190,12 @@ module.exports = grammar({
...
+
     _non_special_token: $ => choice(
       $._literal, $.identifier, $.metavariable, $.mutable_specifier, $.self, $.super, $.crate,
       alias(choice(...primitive_types), $.primitive_type),
-      /[/_\-=->,;:::!=?.@*&#%^+<>|~]+/,
+      $.non_special_punctuation,
```
Comment (maxbrunsfeld) [inline grammar.js:198]: I think that if we're going to make this visible, it would be better to parse these tokens the same as when they appear in other contexts. I believe that in Rust macros, the token trees need to consist of valid Rust tokens. So we don't need a separate, catch-all regex for handling all of these. Instead, we should just list out all of the valid punctuation tokens. They can be found in [the language reference](https://doc.rust-lang.org/reference/tokens.html#punctuation).

```js
{
  _non_special_token: $ => choice(
    // named tokens...
    // keywords...
    '+', '-', '*', '/', '%', '^', '!',  // all of the other punctuation tokens...
  )
}
```

This will actually simplify the parser, since we'll be reusing existing tokens and can get rid of this catch-all regex completely.

Comment (maxbrunsfeld) [reply grammar.js:198]: Once you do this, the corpus files will require much smaller changes, since these tokens will be anonymous nodes, just like they are in other contexts.

**Comment by aryx** [comment]
BTW @Wilfred did you receive the email I've sent you separately? (from pad@r2c.dev)

File: corpus/macros.txt (hunk 1 above)
Comment (Wilfred) [reply corpus/macros.txt]: Argh, that wasn't intentional. Something about my setup (maybe a differing tree-sitter version?) changed this. I'll fix.

**Comment by Wilfred** [comment]
OK, I've replaced the regex with an explicit list of tokens. I've kept the named node, so it's available in the parse tree.

Please take a look and let me know what you think :)

**Comment by amaanq** [comment]
closing as stale
